pulp_container/app/models.py

```diff
@@ -487,6 +547,13 @@ class ContainerRepository(
         ManifestSigningService, on_delete=models.SET_NULL, null=True
     )
 
+    # temporary relations used for uncommitted pull-through cache operations
+    pending_tags = models.ManyToManyField(Tag)
+    pending_manifests = models.ManyToManyField(Manifest)
+    pending_blobs = models.ManyToManyField(Blob, related_name="pending_blobs")
+    # digests of remaining blobs to be attached to pending manifests
+    remaining_blobs = models.ManyToManyField(Blob, related_name="remaining_blobs")
```

Comment:
While downloading/caching layers from a manifest, we track how many of them were already downloaded and how many are still missing. Once there are no remaining blobs to download, we commit the whole box of content, consisting of a tag, manifest, and blobs, to a new repository version: https://github.com/lubosmj/pulp_container/blob/f6e057f3ae5c44868155cb5657daf072b24f9ec4/pulp_container/app/registry_api.py#L982.

This relation was introduced to omit file access to the manifest's artifact each time an associated blob is downloaded. We read the manifest's artifact once at the beginning of the pull-through caching and then record all blobs that will be downloaded next (tag -> manifest -> blob1,blob2,blob3). During the download, we do not have a manifest object containing all FK references to layers/blobs. We build this relation just before committing the content to the repository, when everything is downloaded in place.

If we make the Manifest model artifactless, the relation will be redundant.

Reply:
We are not going to make manifests artifact-less, are we?
Maybe getting ahead of myself now, but do we need ephemeral manifests that blobs can link to, but that cannot be added to repositories before being converted into real ones? (A bit like how an Upload is an ephemeral/incomplete Artifact.)

Reply:
We are going to make the manifests artifact-less soon-ish. By using real "ephemeral" manifests, we re-use what is already provided, right? The problem could be the migration that should be managed without downtime. Do you think having the `ManifestUpload` model is the way to go?

Reply:
oh, artifactless is different than Blob less, right?
Can we still state that a Manifest can only be added to the database with all the blob relations it will ever have in the same transaction?

Reply:
I believe we cannot do that. The sync pipeline used to do it and we got into a lot of trouble with failed, incomplete syncs. Adding a blob to a manifest is like modifying it (while we guarantee unmutability).

The scenario i know that can go wrong is: A user adds an unfinished Manifest to their repository, and the task will carry only the by then associated blobs with it, leading to an invalid repository version. And the best solution I see is to prevent unfinished Manifests to exist in the first place.